buzzmobile/tests/test_utils/mock_nodes.py

```diff
@@ -4,114 +4,138 @@
         """Sends data to be publoshed to the mocked topic.
         """
-        self.node.publish(value)
+        value.serialize(self.proc.stdin)
         # Just long enough to prevent out of order
-        time.sleep(.1)
+        time.sleep(.2)
+
+    def kill(self):
+        """Kills the publisher ros node process.
+        """
+        self.proc.kill()
+
 
 @contextlib.contextmanager
 def mock_pub(topic, rosmsg_type, queue_size=1):
     """Mocks a node and cleans it up when done.
     """
-    pub = rospy.Publisher(topic, rosmsg_type, queue_size=queue_size)
-    yield MockNode(topic, rosmsg_type, pub)
-    pub.unregister()
+    pub = MockPublisher(topic, rosmsg_type, queue_size)
+    no_ns = topic.split('/')[-1]
+    while not any(nn.split('/')[-1].startswith(
+        ''.join(['mock_publish_', no_ns])) for nn in rosnode.get_node_names()):
+        time.sleep(.1)
+    try:
+        yield pub
+    finally:
+        pub.kill()
 
 
-class TestNode(object):
+class MockListener(object):
     """Wrapper around a node used for testing.
     """
 
-    def __init__(self, topic, msg_type):
+    def __init__(self, topic, msg_type, timeout=10):
+        self.timeout = timeout
         self.topic = topic
         self.msg_type = msg_type
-        self.received = False
+        self.killed = False
+
+        # this might be fragile if not run from the ci_scripts/unittest command
+        location = './tests/test_utils/listener.py'
+        self.proc = subprocess.Popen([location,
+            pickle.dumps((topic, msg_type))], stdout=subprocess.PIPE)
         self._message = None
 
+    def kill(self):
+        """Kills the subscriber ros node process.
+        """
+        self.proc.kill()
+        self.proc.wait()
+
     @property
     def message(self):
         """Getter for the message property.
 
         Makes sure you've actually recieved a message before providing one.
         """
-        if not self.received:
-            raise NoMessage(('No message has been '
-                'published to {} yet').format(self.topic))
+        if  not self._message:
+            msg = self.msg_type()
+            sio = StringIO()
+            msg.serialize(sio)
+            # This weird business with StringIO is to make sure we read the
+            # correct number of bytes. Essentially, a serialized ros message
+            # is a fixed length string sent over the channel, so we serialize
+            # the default message of a type and then read that number of bytes
+            # so that we don't encounter any issues with reading too much or
+            # too little data.
+
+            def kill_proc():
+                """Kills the proccess and sets a flag to raise an error.
+                """
+                self.proc.kill()
+                self.killed = True
+
+            timer = threading.Timer(self.timeout, kill_proc)
+            try:
+                timer.start()
+                data = self.proc.stdout.read(sio.len)
```

Comment:
quick clarification question:

if i start a publisher node, start a listener node, publish a message, and do listener.message *after* that message was sent, will this still capture that? I assume so bc stdin will wait until its consumed, but I just want to make sure that the following would still work:

```python
def test(self):
    with mock_pub('/some') as p:
        with check_topic('/some') as ct:
            p.publish()
            time.sleep(1000)
            assert ct.message
```